### ISSUE TITLE
Discount creation error

### ISSUE THREAD

**Issue by chrisvidal** [OFFLINE-GmbH/oc-mall-plugin#377]
cant create any discount since last version 

`Allowed memory size of 134217728 bytes exhausted (tried to allocate 20480 bytes)`
`/var/www/html/vendor/october/rain/src/Database/Traits/Nullable.php line 27`

**Comment by chrisvidal** [comment]
i am trying to figure out where it could come from.

what I noticed in my database in prod:
- they are 65 000 cart rows
- there are 10 cart_discount rows that

i wonder if the memory error is coming from here.

When we load a discount, it loads all the existing carts (belongToMany) and crashes

**Comment by chrisvidal** [comment]
should we have an expiry cart schedule job to clear the cart table?

**Comment by tobias-kuendig** [comment]
At the moment, this can be done with a scheduled job, yes. We will optimize the cart creation in a future release so that a cart is only created once the user added something to it.

Does the out of memory error happen in the backend? Is the above line everything from the stack trace?

**Comment by chrisvidal** [comment]
yes exception is coming from the backend.

they are 4 discounts.

1 can be edited without problem.
3 generates the exception

and create new one is creating exception as well.

rest of line:

```
throw new Exception(sprintf(
                'You must define a $nullable property in %s to use the Nullable trait.', get_called_class()
            ));
        }
  
        static::extend(function($model) {
            $model->bindEvent('model.beforeSave', function() use ($model) {
                $model->nullableBeforeSave();
            });
        });
    }
```

**Comment by chrisvidal** [comment]
![Screenshot 2019-11-14 14 40 54](https://user-images.githubusercontent.com/2263496/68832766-c7103c80-06ec-11ea-8bba-9f3c5051ec98.png)

**Comment by chrisvidal** [comment]
I am keen to develop a cart cleaner job ~ any idea about the request to get the list of carts to be deleted and make sure we don't delete any cart still "usefull" ?

**Comment by tobias-kuendig** [comment]
```php
OFFLINE\Mall\Models\Cart
    ::withTrashed()
    ->whereDoesntHave('customer')
    ->where('updated_at', '<', now()->subDays(7))
    ->get()
    ->each
    ->forceDelete();
```

If you've got a large number of carts, you might need to use `chunk` instead of `get`. Ajdust the updated_at deadline as you wish.

Make sure to call `forceDelete` on every model separately so the attached `CartProduct` models are removed as well. This is done using the `each` call in the example above. Also make sure to use `forceDelete` since the cart models are only soft-deleted otherwise.

**Comment by chrisvidal** [comment]
very interesting, I was surprise at first by the results returned by the query and did some testing:

```php
      $q = \OFFLINE\Mall\Models\Cart
            ::withTrashed()
            ->whereDoesntHave('customer')
            ->where('updated_at', '<', now()->subDays(7))
            ;
```
returns 56144  results from the sql:

```sql
select * from `offline_mall_carts` where not exists (select * from `offline_mall_customers` where `offline_mall_carts`.`customer_id` = `offline_mall_customers`.`id` and `offline_mall_customers`.`deleted_at` is null) and `updated_at` < ?  
```

```php

        $q = \OFFLINE\Mall\Models\Cart
            ::withTrashed()
            ->where('customer_id', null)
            ->where('updated_at', '<', now()->subDays(7))
            ;
```

returns 55948 rows from the sql:

```sql
select * from `offline_mall_carts` where `customer_id` is null and `updated_at` < ?  
```


in the second case, I notice that the `withTrashed` is ignored in the sql.

the sql coming from the 1st query is quite interesting and different from what I would have expected. the difference in results means that customers have been deleted, but their cart remained.

**Comment by tobias-kuendig** [comment]
That's strange. It looks like the `withTrashed` scope is not applied at all:
```php
\OFFLINE\Mall\Models\Cart::withTrashed()->toSql();

=> "select * from `offline_mall_carts`"
```

**Comment by tobias-kuendig** [comment]
Scratch that, the `withTrashed()` scope abviously *removes* the `where deleted_at is null` condition.

In your example above, query #1 returns all old carts for existing customers, but not carts for deleted customers (the Customer model is soft deletable too). #2 returns all carts that don't have a customer attached at all.

I guess #2 is what you want in this case.

**Comment by chrisvidal** [comment]
ya, I am using the query 2 to make a good cleaning :)

but we still got the `memory error` on the backend when trying to create a new discount, with a remaining 14 000 carts. not much more trace.
got the same memory error when I search the word 'international', any other search works fine.
this word 'international' is used a lot in the 15000 products/variants in the project

**Comment by chrisvidal** [comment]
just leaving a note for myself:
I think the `memory` error is coming from the 'product' relation in the discount model.
in form field, it is handled through a dropdown, and I have more than 15000 items. I think it would be better to use a recordfinder here. I will try and let you know later.

**Comment by chrisvidal** [comment]
ok, with the record funder, the memory error disappears:

in `/plugins/offline/mall/models/discount/fields.yaml`

at line 87

```yaml
   product:
        label: 'offline.mall::lang.common.product'
        oc.commentPosition: ''
        descriptionFrom: description
        span: auto
        type: recordfinder
        list: ~/plugins/offline/mall/models/product/columns.yaml
        prompt: Click the %s button to find a user
        nameFrom: name
        trigger:
            action: show
            field: trigger
            condition: value[product]
```

**Comment by tobias-kuendig** [comment]
Yes, I believe there was another user that had this problem once where many products (or categories?) were present. I guess it makes sense to switch to a record finder here.

**Comment by tobias-kuendig** [comment]
This is fixed in 0f66b70, many thanks @chrisvidal!
